src/include/intarith.h

```diff
@@ -220,4 +221,62 @@ template<class T>
   return (sizeof(v) * 8) - __builtin_clzll(v);
 }
 
+namespace ceph {
+namespace math {
+
+template<class UnsignedValueT>
+class p2_t {
+  typedef uint8_t exponent_type;
+  typedef UnsignedValueT value_type;
+
+  static_assert(std::is_unsigned_v<value_type>);
+  static_assert(std::numeric_limits<exponent_type>::max() >
+		std::numeric_limits<value_type>::digits);
+
+  value_type value;
+
+  struct _check_skipper_t {};
+  p2_t(const value_type value, _check_skipper_t)
+    : value(value) {
+  }
+
+public:
+  p2_t(const value_type value)
+    : value(value) {
+    // 0 isn't a power of two. Additional validation is necessary as
+    // the isp2 routine doesn't sanitize that case.
+    //assert(value != 0);
```

Comment:
Hmm, I have mixed feelings here. Our `isp2` and `ISP2` routines treat `0` as a power of `2`. We can just live with that or be pedantic and rename everything to e.g. `isp2o0` and `pt2o0_t`. Definitely I would like to avoid confusion coming from distinct naming conventions.

Also, having the ability to store `0` in `p2_t` looks very appealing as the existing code puts `0` into things commented as holding `p2` only.